### ISSUE TITLE
Consider refactoring of Compiler::numArgsOfHWIntrinsic function to improve performance

### ISSUE THREAD

**Issue by 4creators** [dotnet/coreclr#16744]
See discussion from this comment on: 

https://github.com/dotnet/coreclr/pull/16736#issuecomment-370188525

**Comment by CarolEidt** [comment]
Before we spend any time on this, I would like to see some evidence that this actually matters.

**Comment by mikedn** [comment]
> Before we spend any time on this, I would like to see some evidence that this actually matters.

Yep, this whole story is rather speculative. Intrinsic usage will always be pretty low compared to "normal" code so it will be difficult to measure the impact of such a change. It's probably good but too small to matter, especially when comparing it with other zillion of perf issues the JIT has.

Not to mention that coreclr uses PGO and that has a significant impact on inlining and since there's no intrinsic usage PGO may very well decide not to inline even the "improved" function...

**Comment by CarolEidt** [comment]
In looking at the code, one refactoring that might actually be useful would be to factor out the code that actually counts the node's operands, e.g. `countArgsOfHWIntrinsic(GenTreeHWIntrinsic* node)`. Then, in the case where the `numArgs` field is not `-1`, you could add an assert that it returns the same value as `numArgs`. I suspect it might be useful for asserts elsewhere as well.

**Comment by CarolEidt** [comment]
> Not to mention that coreclr uses PGO and that has a significant impact on inlining and since there's no intrinsic usage PGO may very well decide not to inline even the "improved" function...

Yes, and PGO is a bit tricky because unless you re-run the measurement pass, making changes to the source and rebuilding wont' actually show you what the optimizer will do.

**Comment by 4creators** [comment]
> one refactoring that might actually be useful would be to factor out the code that actually counts the node's operands, e.g. countArgsOfHWIntrinsic(GenTreeHWIntrinsic* node). 

This is exactly what I wanted to test.

> Yes, and PGO is a bit tricky because unless you re-run the measurement pass, making changes to the source and rebuilding wont' actually show you what the optimizer will do.

Yep, that's the problem.

> Before we spend any time on this, I would like to see some evidence that this actually matters.

Agree, will run some profiles to check the impact and potential savings
